src/runtime/process.cpp

```diff
@@ -160,19 +160,41 @@ static obj_res spawn(string_ref const & proc_name, array_ref<string_ref> const &
     siStartInfo.hStdError = child_stderr;
     siStartInfo.dwFlags |= STARTF_USESTDHANDLES;
 
-    // TODO(gabriel): this is thread-unsafe
-    std::unordered_map<std::string, optional<std::string>> old_env_vars;
-    for (auto & entry : env) {
-        optional<std::string> old;
-        if (auto old_val = getenv(entry.fst().data()))
-            old = std::string(old_val);
-        old_env_vars[entry.fst().to_std_string()] = old;
-
-        if (entry.snd()) {
-            SetEnvironmentVariable(entry.fst().data(), entry.snd().get()->data());
-        } else {
-            SetEnvironmentVariable(entry.fst().data(), nullptr);
+    char * new_env = nullptr, * esp = nullptr;
+    std::unique_ptr<char> new_env_buf;
+
+    if (env.size()) {
+        std::unordered_map<std::string, std::string> new_env_vars; // C++17 gives us no-copy std::string_view for this, much better!
+        for (auto & entry : env) {
+            new_env_vars[entry.fst().data()] = entry.snd() ? entry.snd().get()->data() : std::string{};
         }
+        static constexpr auto env_buf_size = 0x8000;
+        new_env = new char[env_buf_size];
+        new_env_buf = std::unique_ptr<char>(new_env);
+        esp = GetEnvironmentStrings();
+        // First copy old evars not in new evars.
+        char *nep = new_env, *ep = esp, *oep = ep;
+        while (*ep) {
+            char *eqp = ep = strchr(ep, '=');
+            ep += strlen(ep);
+            if (!new_env_vars.count({oep, static_cast<size_t>(eqp - oep)})) {
+                nep = std::copy(oep, ep, nep);
+                *nep++ = '\0';
+            }
+            ep++;
+            oep = ep;
+        }
```

Comment:
```suggestion
        for (char *entryp = esp; *ep;) {
            char *equalsp = strchr(entryp, '=');
            size_t entry_len = strlen(entryp);
            if (!new_env_vars.count({entryp, static_cast<size_t>(equalsp - entryp)})) {
                std::copy(entryp, entryp + entry_len + 1, std::back_inserter(new_env));
            }
            entryp += entry_len;
        }
```
Please try to use (easily) decipherable variable names, scope them as locally as possible, and avoid mutating them during the loop body.

Reply:
Please also declare the `esp` inside the if-block (and free it here as well).